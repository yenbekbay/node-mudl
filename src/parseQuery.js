/* @flow */

import _ from 'lodash/fp';
import he from 'he';

export const featuringArtistRegex = /\(?feat\. ([^()]*)\)?/;

export type TrackQuery = {
  primaryArtist: string,
  artist: string,
  title: string,
  version: ?string
};

export default (...args: Array<string>): ?TrackQuery => {
  let components = _.flow(
    _.replace(/\s+(‒|–|—)\s+/g, ' - '),
    _.split(' - '),
    _.map(_.trim),
  )(he.decode(args.join(' - ')));

  if (components.length < 2) return null;

  components = [components[0], _.drop(1)(components).join(' - ')];

  // Capitalize each word in title if query doesn't contain any foreign chars
  if (!/[^\x00-\x7F]+/.test(components.join(' - '))) {
    components = [
      components[0],
      components[1].replace(/\b\w/g, _.toUpper),
    ];
  }

  components = _.map((comp: string): string => comp
    .replace(/\//g, ' & ')
    .replace(/(\s+|\()f((eat|t)\.?|eaturing)\s+/ig, '$1feat. ')
    .replace(/\s+v(s\.?|ersus)\s+/ig, ' vs. ')
    .replace(/\s+dj\s+/ig, ' DJ ')
    .replace(/\[/g, '(')
    .replace(/\]/g, ')')
    .replace(/\(/g, ' ('),
  )(components);

  const getComponent = (idx: number): string => components[idx]
    .replace(featuringArtistRegex, '')
    .replace(/\s+/g, ' ')
    .trim();

  const featuringArtists = _.flow(
    _.map((comp: string): string => _.flow(
      _.nth(1),
      _.trim,
    )(comp.match(featuringArtistRegex))),
    _.compact,
  )(components);
  const primaryArtist = getComponent(0);
  const artist = primaryArtist + (
    featuringArtists && featuringArtists.length
      ? ` feat. ${featuringArtists.join(' & ')}`.replace(/[Aa]nd/g, '&')
      : ''
  );
  const title = getComponent(1)
    .replace(
      /\([^\(]*(download|preview|out now|premiere|recordings)[^\)]*\)/ig,
      '',
    )
    .trim();
  const version = _.nth(1)(title.match(/((?:\([^\(]*\)\s?)+$)/));

  return { primaryArtist, artist, title, version };
};
