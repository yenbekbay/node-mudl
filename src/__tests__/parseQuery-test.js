/* @flow */

import parseQuery from '../parseQuery';

const featVariants = ['ft', 'ft.', 'feat', 'feat.', 'Feat', 'Feat.'];
const promoTexts = [
  'Download', 'Preview', 'OUT NOW', 'Premiere', 'Something Recordings',
];

const testQueries = [
  'Some Artist - Some Track',
  'Some Artist - Some Track (Some Remix)',
  ...featVariants.map(
    (feat: string) => `Some Artist ${feat} Another Artist - Some Track`,
  ),
  ...featVariants.map(
    (feat: string) => `Some Artist - Some Track (${feat} Another Artist)`,
  ),
  'some artist - some track (some remix)',
  'Какой-то артист - Какой-то трек (Какой-то ремикс)',
  ...promoTexts.map(
    (promo: string) => `Some Artist - Some Title (Some Remix) [${promo}]`,
  ),
];

describe('parseQuery', () => {
  it('processes a query', () => {
    testQueries.forEach((query: string) => {
      expect(parseQuery(query)).toMatchSnapshot();
    });
  });
});
