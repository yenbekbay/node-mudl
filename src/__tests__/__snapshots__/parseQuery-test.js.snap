exports[`parseQuery processes a query 1`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 2`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 3`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 4`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 5`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 6`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 7`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 8`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 9`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 10`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 11`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 12`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 13`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 14`] = `
Object {
  "artist": "Some Artist feat. Another Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track",
  "version": undefined,
}
`;

exports[`parseQuery processes a query 15`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Track (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 16`] = `
Object {
  "artist": "Какой-то артист",
  "primaryArtist": "Какой-то артист",
  "title": "Какой-то трек (Какой-то ремикс)",
  "version": "(Какой-то ремикс)",
}
`;

exports[`parseQuery processes a query 17`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Title (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 18`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Title (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 19`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Title (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 20`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Title (Some Remix)",
  "version": "(Some Remix)",
}
`;

exports[`parseQuery processes a query 21`] = `
Object {
  "artist": "Some Artist",
  "primaryArtist": "Some Artist",
  "title": "Some Title (Some Remix)",
  "version": "(Some Remix)",
}
`;
